sim/core/item_swaps.go

```diff
@@ -136,80 +191,180 @@ func (swap *ItemSwap) RegisterOnSwapItemForEnchantEffect(effectID int32, aura *A
+	if !swap.IsEnabled() || swap.swapSet == swapSet {
 		return
 	}
 
 	character := swap.character
 
 	meleeWeaponSwapped := false
 	newStats := stats.Stats{}
-	has2H := swap.GetItem(proto.ItemSlot_ItemSlotMainHand).HandType == proto.HandType_HandTypeTwoHand
+	has2H := swap.GetUnequippedItemBySlot(proto.ItemSlot_ItemSlotMainHand).HandType == proto.HandType_HandTypeTwoHand
+	isPrepull := sim.CurrentTime < 0
 	for _, slot := range slots {
+		if !isReset && !isPrepull && (slot < proto.ItemSlot_ItemSlotMainHand || slot > proto.ItemSlot_ItemSlotRanged) {
+			continue
+		}
+
 		//will swap both on the MainHand Slot for 2H.
 		if slot == proto.ItemSlot_ItemSlotOffHand && has2H {
 			continue
 		}
 
-		if ok, swapStats := swap.swapItem(slot, has2H); ok {
+		if ok, swapStats := swap.swapItem(slot, has2H, isReset); ok {
 			newStats = newStats.Add(swapStats)
-			meleeWeaponSwapped = slot == proto.ItemSlot_ItemSlotMainHand || slot == proto.ItemSlot_ItemSlotOffHand || meleeWeaponSwapped
+			meleeWeaponSwapped = slot == proto.ItemSlot_ItemSlotMainHand || slot == proto.ItemSlot_ItemSlotOffHand || slot == proto.ItemSlot_ItemSlotRanged || meleeWeaponSwapped
+
+			for _, onSwap := range swap.onSwapCallbacks[slot] {
+				onSwap(sim, slot)
+			}
 		}
 	}
 
 	if sim.Log != nil {
-		sim.Log("Item Swap Stats: %v", newStats.FlatString())
+		sim.Log("Item Swap - New Stats: %v", newStats.FlatString())
 	}
-
 	character.AddStatsDynamic(sim, newStats)
 
-	for _, onSwap := range swap.onSwapCallbacks {
-		onSwap(sim)
-	}
-
-	if character.AutoAttacks.AutoSwingMelee && meleeWeaponSwapped && sim.CurrentTime > 0 {
-		character.AutoAttacks.StopMeleeUntil(sim, sim.CurrentTime, false)
-	}
+	if !isPrepull && !isReset {
+		if character.AutoAttacks.AutoSwingMelee && meleeWeaponSwapped {
+			character.AutoAttacks.StopMeleeUntil(sim, sim.CurrentTime, false)
+		}
 
-	// If GCD is ready then use the GCD, otherwise we assume it's being used along side a spell.
-	if character.GCD.IsReady(sim) {
-		newGCD := sim.CurrentTime + 1500*time.Millisecond
-		character.SetGCDTimer(sim, newGCD)
+		// If GCD is ready then use the GCD, otherwise we assume it's being used along side a spell.
+		if character.GCD.IsReady(sim) {
+			newGCD := sim.CurrentTime + 1500*time.Millisecond
+			character.SetGCDTimer(sim, newGCD)
+		}
 	}
 
-	swap.swapped = !swap.swapped
+	swap.swapSet = swapSet
 }
 
-func (swap *ItemSwap) swapItem(slot proto.ItemSlot, has2H bool) (bool, stats.Stats) {
-	oldItem := swap.character.Equipment[slot]
-	newItem := swap.GetItem(slot)
+func (swap *ItemSwap) swapItem(slot proto.ItemSlot, has2H bool, isReset bool) (bool, stats.Stats) {
+	oldItem := *swap.GetEquippedItemBySlot(slot)
+	var newItem *Item
+	if isReset {
+		newItem = &swap.originalEquip[slot]
+	} else {
+		newItem = swap.GetUnequippedItemBySlot(slot)
+	}
 
 	swap.character.Equipment[slot] = *newItem
 	oldItemStats := swap.getItemStats(oldItem)
```

Comment:
You can probably eliminate a lot of this pointer magic if you have `getItemStats()` take in *Item instead of Item?

Reply:
Related question: if the goal of using pointers here is performance-driven to avoid copying `Item` structs, then why not refactor `Equipment` itself to be an array of pointers instead? That way you can just directly access the equipped and unequipped arrays by slot index without the need for these helper access methods.

Reply:
I am not sure how to best tackle this. I just followed and extended on the initial implementation. Refactoring the Equipment itself is outside of my expertise